cluster/addons/dns/nodelocaldns/nodelocaldns.yaml.base

```diff
@@ -0,0 +1,152 @@
+# Copyright 2018 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# __MACHINE_GENERATED_WARNING__
+
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: nodelocaldns
+  namespace: kube-system
+  labels:
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: Reconcile
+---
+
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: nodelocaldns
+  namespace: kube-system
+  labels:
+    addonmanager.kubernetes.io/mode: EnsureExists
+
+data:
+  Corefile: |
+    __PILLAR__DNS__DOMAIN__:53 {
+        errors
+        cache 30
+        reload
+        loop
+        bind __PILLAR__LOCAL__DNS__
+        forward . __PILLAR__DNS__SERVER__ {
+                force_tcp
+        }
+        prometheus :9253
+        health __PILLAR__LOCAL__DNS__:8080
+        }
+    in-addr.arpa:53 {
+        errors
+        cache 30
+        reload
+        loop
+        bind __PILLAR__LOCAL__DNS__
+        forward . __PILLAR__DNS__SERVER__ {
+                force_tcp
+        }
+        prometheus :9253
+        }
+    ip6.arpa:53 {
+        errors
+        cache 30
+        reload
+        loop
+        bind __PILLAR__LOCAL__DNS__
+        forward . __PILLAR__DNS__SERVER__ {
+                force_tcp
+        }
+        prometheus :9253
+        }
+    .:53 {
+        errors
+        cache 30
+        reload
+        loop
+        bind __PILLAR__LOCAL__DNS__
+        forward . /etc/resolv.conf {
+                force_tcp
+        }
+        prometheus :9253
+        }
+---
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: nodelocaldns
+  namespace: kube-system
+  labels:
+    k8s-app: kube-dns
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: Reconcile
+spec:
+  selector:
+    matchLabels:
+      k8s-app: nodelocaldns
+  template:
+    metadata:
+       labels:
+          k8s-app: nodelocaldns
+    spec:
+      priorityClassName: system-node-critical
+      serviceAccountName: nodelocaldns
+      hostNetwork: true
+      dnsPolicy: Default  # Don't use cluster DNS.
+      nodeSelector:
+        beta.kubernetes.io/nodelocaldns-ready: "true"
```

Comment:
What does this label mean, and who sets it?

Do we really need it?

Reply:
We need it so that the cache does not run on master node. This label is set in the config files at cluster startup as a NON_MASTER_NODE_LABEL
I saw that other kube-system daemonsets - flutentd, metadata proxy have similar *-ready labels assigned to nodes and master and use that as a selector.

Reply:
> I saw that other kube-system daemonsets - flutentd, metadata proxy have similar *-ready labels assigned to nodes and master and use that as a selector.

those labels are problematic for several reasons:
* they only get added on node creation, not on node upgrade
* they depend on the kubelet being able to self-label arbitrarily (close-to-reaching-agreement design restricting that in https://github.com/kubernetes/community/pull/911)
* they promote the model of nodes essentially requesting specific workloads, rather than workloads selecting nodes based on fundamental characteristics of the nodes themselves

Reply:
I actually thought the `-ready` nodes was more about controlling rollout?  For the case of fluentd, I thought it was so that we could rolling-update from a static pod (IIRC) to a daemonset.

If we're just copying fluentd (I'm not sure about metadata proxy), maybe we don't need it.
 
Not that I don't love a good master-label debate, but I don't think we have to have it here :-)

Reply:
Removed the label.

Reply:
Since this requires a kubelet command line reconfiguration, can we go without a node version gate?